src/corecel/sys/ScopedDeviceProfiling.cuda.cc

```diff
@@ -22,20 +27,75 @@ namespace celeritas
 {
 namespace
 {
+
+//---------------------------------------------------------------------------//
+/*!
+ * Global registry for strings used by NVTX.
+ * This Implemented as a free function instead of a static in
+ * ScopedDeviceProfiling to hide the NVTX dependency from user of the
+ * interface.
+ */
+std::unordered_map<std::string, nvtxStringHandle_t>& message_registry()
+{
+    static std::unordered_map<std::string, nvtxStringHandle_t> registry;
+    return registry;
+}
+
 nvtxDomainHandle_t domain_handle()
 {
     static nvtxDomainHandle_t domain = nvtxDomainCreateA("celeritas");
     return domain;
 }
+
+//---------------------------------------------------------------------------//
+/*!
+ * Retrieve the handle for a given message, insert it if it doesn't already
+ * exists
+ */
+nvtxStringHandle_t message_handle_for(std::string const& message)
+{
+    static std::shared_mutex mutex;
+
+    {
+        std::shared_lock lock(mutex);
+        auto message_handle = message_registry().find(message);
+        if (message_handle != message_registry().end())
+        {
+            return message_handle->second;
+        }
+    }
+    // we did not find the handle, insert it
+    std::unique_lock lock(mutex);
+    auto message_handle = message_registry().find(message);
+    // recheck that nobody inserted the same message as we had to release the
+    // shared lock not strictly necessary as insert will do nothing if the key
+    // already exists
+    if (message_handle == message_registry().end())
+    {
+        auto handle
+            = nvtxDomainRegisterStringA(domain_handle(), message.c_str());
+        message_registry().insert({message, handle});
+        return handle;
+    }
+    else
+    {
+        return message_handle->second;
+    }
```

Comment:
The solution is to just `insert` with a `nullptr` for the handle, and if insertion succededed to register it and save in the returned iterator.
```suggestion
    auto [iter, inserted] = message_registry().insert({message, nullptr});
    if (inserted)
    {
        iter->second
            = nvtxDomainRegisterStringA(domain_handle(), message.c_str());
    }
    CELER_ENSURE(iter->second);
    return iter->second;
```

Reply:
@sethrj Actually, we can't really `CELER_ENSURE` that iter->second is non-null. If we're not running the application through a tool that can handle nvtx, i.e. ncu with --nvtx opthon, then all API calls will return 0 ([ref](https://github.com/NVIDIA/NVTX#how-much-overhead-does-nvtx-add-to-my-code)).

Also we can't really make any assumption on what values are returned. I tried having multiple nested ranges and printing each value returned by `nvtxDomainRegisterStringA`. The first call returns 0, the second call returns 1 so they're not really pointers as you'd expect ([relevant doc](https://github.com/NVIDIA/NVTX#give-dont-take)).

TLDR; just don't call CELER_ENSURE and trust the implementation?

Reply:
Oh interesting. I guess "handle" doesn't mean "pointer" in this case. (Maybe replace `nullptr` with `{}`?) That's really good to know. Could you add a note in the implementation that the API calls may fail if the profiler is disabled?